File: Sources/Shared/Entity/Statics/EntityRarity.ts
import type { ColorCode } from "../../Utils/Color";

export const enum Rarity {
    COMMON,
    UNUSUAL,
    RARE,
    EPIC,
    LEGENDARY,
    MYTHIC,
    ULTRA,
}

export const kMaxRarities = 7 as const;

type RarityDisplayName = Capitalize<string>;

export const RARITY_DISPLAY_NAME = {
    [Rarity.COMMON]: "Common",
    [Rarity.UNUSUAL]: "Unusual",
    [Rarity.RARE]: "Rare",
    [Rarity.EPIC]: "Epic",
    [Rarity.LEGENDARY]: "Legendary",
    [Rarity.MYTHIC]: "Mythic",
    [Rarity.ULTRA]: "Ultra",
} as const satisfies Record<Rarity, RarityDisplayName>;

export const RARITY_COLOR = {
    [Rarity.COMMON]: "#7eef6d",  
    [Rarity.UNUSUAL]: "#ffe65d", 
    [Rarity.RARE]: "#4d52e3",    
    [Rarity.EPIC]: "#861fde",    
    [Rarity.LEGENDARY]: "#de1f1f",
    [Rarity.MYTHIC]: "#1fdbde",  
    [Rarity.ULTRA]: "#ff2b75", 
} as const satisfies Record<Rarity, ColorCode>;

File: Sources/Client/private/src/Florr/Native/Entity/EntityRarity.zig
const std = @import("std");
const meta = std.meta;
const Color = @import("../../../WebAssembly/Interop/Canvas/Color.zig");

pub const EntityRarity = enum(u8) {
    common,
    unusual,
    rare,
    epic,
    legendary,
    mythic,
    ultra,
};

const rarity_fields = meta.fields(EntityRarity);

pub const max_rarities: u8 = rarity_fields[rarity_fields.len - 1].value;

const rarity_names = std.StaticStringMap([]const u8).initComptime(.{
    .{ @tagName(EntityRarity.common), "Common" },
    .{ @tagName(EntityRarity.unusual), "Unusual" },
    .{ @tagName(EntityRarity.rare), "Rare" },
    .{ @tagName(EntityRarity.epic), "Epic" },
    .{ @tagName(EntityRarity.legendary), "Legendary" },
    .{ @tagName(EntityRarity.mythic), "Mythic" },
    .{ @tagName(EntityRarity.ultra), "Ultra" },
});

const rarity_colors = std.StaticStringMap(Color).initComptime(.{
    .{ @tagName(EntityRarity.common), Color.comptimeFromHexColorCode("#7EEF6D") },
    .{ @tagName(EntityRarity.unusual), Color.comptimeFromHexColorCode("#FFE65D") },
    .{ @tagName(EntityRarity.rare), Color.comptimeFromHexColorCode("#4D52E3") },
    .{ @tagName(EntityRarity.epic), Color.comptimeFromHexColorCode("#861FDE") },
    .{ @tagName(EntityRarity.legendary), Color.comptimeFromHexColorCode("#DE1F1F") },
    .{ @tagName(EntityRarity.mythic), Color.comptimeFromHexColorCode("#1FDBDE") },
    .{ @tagName(EntityRarity.ultra), Color.comptimeFromHexColorCode("#373737") },
});

pub inline fn getRarityName(rarity: EntityRarity) ?[]const u8 {
    return rarity_names.get(@tagName(rarity));
}

pub inline fn getRarityColor(rarity: EntityRarity) ?Color {
    return rarity_colors.get(@tagName(rarity));
}
